### PR TITLE
aften: add test do block

### PR DESCRIPTION
**Pull request by eirinikos**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/homebrew-core/blob/master/.github/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Does your submission pass `brew audit --strict --online <formula>` (where `<formula>` is the name of the formula you're submitting)?
- [x] Have you built your formula locally prior to submission with `brew install <formula>`?
## 

This commit adds a `test do` block to the formula.

### CODE REVIEW

**Comment by tdsmith** [comment]
Thanks; this is great! Is there a chance we could try to actually encode a sample audio file? You can use the `resource` mechanism to download something if the package doesn't install any example files.

**Comment by eirinikos** [comment]
Thanks for pointing me in the direction of the `resource` mechanism! My old job dealt with audio encoding, so it was fun to give this the old college try. It looks like I'm not employing `resource` correctly, though. I pushed a new commit; is it possible for you to take a look and let me know what's wrong?

**Comment by ilovezfs** [comment]
```
==> FAILED
Error: 1 problem in 1 formula
aften:
 * Non-executables were installed to "/usr/local/Cellar/aften/0.0.8/bin"
The offending files are:
  /usr/local/Cellar/aften/0.0.8/bin/1kHz_44100Hz_16bit_05sec.wav
```

Please run `brew audit --strict` and fix the issue(s) it identifies.

In this case you'll probably want to use `share.install` instead of `bin.install`.

File: Formula/aften.rb [hunk 1]
```diff
@@ -13,10 +13,24 @@ class Aften < Formula
...
+    sha256 "949dd8ef74db1793ac6174b8d38b1c8e4c4e10fb3ffe7a15b4941fa0f1fbdc20"
+  end
+
   def install
     mkdir "default" do
       system "cmake", "-DSHARED=ON", "..", *std_cmake_args
       system "make", "install"
     end
+
+    resource("sample_wav").stage { bin.install "1kHz_44100Hz_16bit_05sec.wav" }
```
Comment (tdsmith) [inline Formula/aften.rb]: I would move this to the `test do...` block and write `resource("sample_wav").stage testpath` to install it to the working directory.

File: Formula/aften.rb [hunk 2]
```diff
@@ -13,10 +13,24 @@ class Aften < Formula
...
+  end
+
+  test do
+    assert_match %r{Aften: A\/52 audio encoder}, shell_output("#{bin}/aften -h").lstrip.lines.map(&:chomp).first
+
+    assert_match("output format: 48000 Hz stereo (2/0)",
+      pipe_output("#{bin}/aften 1kHz_44100Hz_16bit_05sec.wav sample.ac3").lines.map(&:chomp).find { |line| line.start_with? "output" })
```
Comment (tdsmith) [inline Formula/aften.rb]: I think just doing:
`system "#{bin}/aften", "#{testpath}/1kHz_44100Hz_16bit_05sec.wav", "sample.ac3"`
is fine (after moving the `resource` staging here into the test block) -- no need to check the output if aften returns a meaningful exit code.

**Comment by tdsmith** [comment]
Thanks for working on this! I left a couple of comments and then this will be great to pull. `Resource#stage` can accept either a block or a path; in the latter format the file is just copied to the path, which is all we need here.

**Comment by eirinikos** [comment]
Thanks for the feedback and explanations, @tdsmith! I'm still getting familiarized with system calls. Everything makes sense so far. I had been thinking that I should employ `testpath`, but I hadn't realized that  the `resource` staging could take place within the test block. I've updated the PR.

Thanks @ilovezfs for pointing that out, I ought to have caught that. For some reason `brew audit --strict` didn't flag any issues. Anyway, it looks like doing the `resource` staging within the test block allows us to avoid that problem.

**Comment by apjanke** [comment]
> Thanks @ilovezfs for pointing that out, I ought to have caught that. For some reason `brew audit --strict` didn't flag any issues. 

Did you run the `brew audit --strict` again after installing the formula locally? Some issues, like installing non-executables or non-libraries, it can only detect once the formula is installed.

Maybe we should modify the issue template to note that the `audit` should be run while the formula is installed; and reorder it so the `install` item comes before the `audit`.

**Comment by DomT4** [comment]
> Maybe we should modify the issue template to note that the audit should be run while the formula is installed; and reorder it so the install item comes before the audit.

This is consistent with the order CI processes every build in so 👍 from me at least.

File: Formula/aften.rb [hunk 3]
```diff
@@ -13,10 +13,22 @@ class Aften < Formula
...
     mkdir "default" do
       system "cmake", "-DSHARED=ON", "..", *std_cmake_args
       system "make", "install"
     end
   end
+
+  test do
+    assert_match %r{Aften: A\/52 audio encoder}, shell_output("#{bin}/aften -h").lstrip.lines.map(&:chomp).first
+
+    resource("sample_wav").stage testpath
```
Comment (DomT4) [inline Formula/aften.rb]: We already have a `.wav` test object available. Can we use that one or is it problematic in this case? The usage would be something like:

```ruby
system "#{bin}/aften", test_fixtures("test.wav"), "sample.ac3"
```

File: Formula/aften.rb [hunk 4]
```diff
@@ -13,10 +13,22 @@ class Aften < Formula
...
+
   def install
     mkdir "default" do
       system "cmake", "-DSHARED=ON", "..", *std_cmake_args
       system "make", "install"
     end
   end
+
+  test do
+    assert_match %r{Aften: A\/52 audio encoder}, shell_output("#{bin}/aften -h").lstrip.lines.map(&:chomp).first
```
Comment (DomT4) [inline Formula/aften.rb]: If I understand what you're trying to do here it's probably cleaner to just do:

```ruby
assert_match "Aften: A 52 audio encoder", shell_output("#{bin}/aften -h 2>&1")
```

File: Formula/aften.rb (hunk 3 above)
Comment (mikemcquaid) [inline Formula/aften.rb]: If that works you can also get rid of the `resource` above. Sorry for not noticing this earlier, @eirinikos!

File: Formula/aften.rb (hunk 4 above)
Comment (mikemcquaid) [inline Formula/aften.rb]: You can also probably remove this test in favour of the better `wav` one below.

**Comment by eirinikos** [comment]
Thanks, everyone!

@DomT4 and @mikemcquaid: I removed the `resource` and the `shell_output` test and tried running the test below....

```
test do
  system "#{bin}/aften", test_fixtures("test.wav"), "test.ac3"
end
```

It appeared to result in a segfault (Segmentation fault: 11, if that means anything). Not sure why. I can't tell if the problem is inherent to the test file, either. How was it generated? What are your thoughts?

@DomT4: `#{bin}/aften -h 2>&1` does look cleaner (for some reason I wrote all that Ruby instead of opting for something like piping to `head`). I'm still fuzzy on what `2>&1` does, but I'll keep it in mind for another time.

**Comment by eirinikos** [comment]
> Did you run the brew audit --strict again after installing the formula locally? Some issues, like installing non-executables or non-libraries, it can only detect once the formula is installed.
> 
> Maybe we should modify the issue template to note that the audit should be run while the formula is installed; and reorder it so the install item comes before the audit.

@apjanke I see, thanks. I had installed the formula prior to editing it... I ought to have reinstalled using the modified formula, then, and then run `brew audit --strict`? This makes sense.

Should [these pull request instructions](https://github.com/Homebrew/brew/blob/master/share/doc/homebrew/How-To-Open-a-Homebrew-Pull-Request-(and-get-it-merged).md#create-your-pull-request-from-a-new-branch) also be updated so that `install` comes before `audit`?

**Comment by apjanke** [comment]
>  I had installed the formula prior to editing it... I ought to have reinstalled using the modified formula, then, and then run brew audit --strict? This makes sense.

Exactly. Because some of those `audit --strict` checks act upon the result of your modified formula's installation instructions.

> Should [these pull request instructions](https://github.com/Homebrew/brew/blob/master/share/doc/homebrew/How-To-Open-a-Homebrew-Pull-Request-(and-get-it-merged).md#create-your-pull-request-from-a-new-branch) also be updated so that `install` comes before `audit`?

Yes, they should. Here's a PR to do so: Homebrew/brew#127. If you have time, could you read through the new language and let us know what you think, and if it would have helped you on this PR? (To view it in a readable form, click on the "Files changed" tab in the PR, and then click on the View button in the upper-right side of the diff view. That'll show you the "after" version of the file, rendered as a web page.)

**Comment by mikemcquaid** [comment]
@eirinikos You could try using the same `.wav` file as before and seeing if it still segfaults. If it doesn't I'd perhaps just go back to using the `resource`. Sorry to pull you back and forth on this!

**Comment by DomT4** [comment]
> I'm still fuzzy on what 2>&1 does, but I'll keep it in mind for another time.

It throws `stderr` and `stdout` into the same output (`stdout`) stream.

**Comment by eirinikos** [comment]
> > I had installed the formula prior to editing it... I ought to have reinstalled using the modified formula, then, and then run brew audit --strict? This makes sense.

> Exactly. Because some of those audit --strict checks act upon the result of your modified formula's installation instructions.
> 

> > Should these pull request instructions also be updated so that install comes before audit?

> Yes, they should. Here's a PR to do so: Homebrew/brew#127. If you have time, could you read through the new language and let us know what you think, and if it would have helped you on this PR? (To view it in a readable form, click on the "Files changed" tab in the PR, and then click on the View button in the upper-right side of the diff view. That'll show you the "after" version of the file, rendered as a web page.)

@apjanke Sorry for the rather late response! I've read through the new language for the PR instructions update and it looks great. I'll make sure to reference the current instructions for future PRs.

One note: if I have already performed a formula installation (prior to changing the formula), I should both `brew remove` and then `brew install` the formula anew (following any changes)... right? I usually already have a formula installed prior to changing it. Running `brew install` again (before removing the formula) just returns a message telling me that it's already installed (if I remember correctly). The current instructions are an improvement, though perhaps it's still slightly unclear that the formula first needs to be removed before re-installing.

I hope that makes sense... please do let me know if there's something I've missed!

**Comment by eirinikos** [comment]
> @eirinikos You could try using the same .wav file as before and seeing if it still segfaults. If it doesn't I'd perhaps just go back to using the resource. Sorry to pull you back and forth on this!

@mikemcquaid No problem and thanks! Now I know where to find the test fixtures for future reference. 😄  I've finally updated the PR and there's no segfault, yay.

Question: I've followed the PR checklist for the changed formula (`brew tests`, `brew install`, `brew test`, and `brew audit --strict`) and while everything checks out, I'm wondering why `:aften.f.test` doesn't work in `brew irb`? I receive:
```
ArgumentError: target directory or block is required
	from /usr/local/Library/Homebrew/resource.rb:82:in `stage'
	from /usr/local/Library/Taps/homebrew/homebrew-core/Formula/aften.rb:29:in `block in <class:Aften>'
	from (irb):2
	from /usr/local/Library/brew.rb:87:in `<main>'
```

Isn't `testpath` recognized as a target directory? Or is it incorrect for me to send `test` to the formula instance?

**Comment by apjanke** [comment]
> One note: if I have already performed a formula installation (prior to changing the formula), I should both brew remove and then brew install the formula anew (following any changes)... right?

Correct. Or do a `brew reinstall <formula>`. Good point; I'll try to update the directions to reflect that.

> wondering why :aften.f.test doesn't work in brew irb?

That `test` method is expecting an argument. Are you supplying one? When you say "`test do`" in a formula, it's not defining a zero-argument `test` method; it's actually passing a block to a previously-defined `test` class method.

See the source for `cmd/test.rb` to find out the right way to invoke a formula's test code from `irb`. (I don't know what it is offhand.)

**Comment by mikemcquaid** [comment]
Thanks again @eirinikos; great work here!

**Pull request merged**